### ISSUE TITLE
FULLSCREEN option: windowed-fullscreen

### ISSUE THREAD

**Issue by aspauldingcode** [koekeishiya/yabai#2221]
Allow a bsp window (or a floating window) to be toggled as fullscreen.

This fullscreen just extends the window to the corners of the display, and toggles ignorance of gaps/bars spacing config to do so.

This way, we can get an alternative to native-fullscreen, which users don't have to script their own gaps/bar toggle respect/ignore keybind to get a bsp window to fullscreen instantly, and it also doesn't require us to write our own toggle for making the window float first and respect the current display dimensions..

**Comment by Bellavene** [comment]
yabai -m window --grid 1:1:0:0:1:1

**Comment by aspauldingcode** [comment]
yabai -m window --grid 1:1:0:0:1:1 || yabai -m window --toggle zoom-fullscreen is already in my setup with alt + f.

However, this obeys my bar and window spacing. 
What I want is a way to simply toggle a complete fullscreen without the native macOS implementation. This will require changes to yabai itself I am sure, otherwise I can create an advanced bash toggle script to disable all my gaps config and bar spacing temperarily to fullscreen the window, but I don't want that to be for the whole space. I wish I could do this for a single window. 

In other words, an instant fullscreen toggle with a command like (yabai -m window --toggle windowed-fullscreen)

**Comment by Bellavene** [comment]
Why native fullscreen doesn't suit you then? If you are using iTerm, then you can use native fullscreen in it without making a space (toggles in the settings.

**Comment by aspauldingcode** [comment]
the issue is still here.
I've created a manual fix but it's slow!
I think if yabai could just toggle fullscreen as an alternative to zoom fullscreen we would have this:
![fullscreenvszoomfullscreen](https://github.com/koekeishiya/yabai/assets/10196826/c81b2cff-d3b5-4d66-952c-1e21d00c7859)

**Comment by aspauldingcode** [comment]
> Why not use native fullscreen for that?

There should be an option that does NOT use macOS's animations. Native Fullscreen fades out, and moves the window to a completely different space.

**Comment by aspauldingcode** [comment]
> I've created a manual fix but it's slow!

This is my second attempt with this.

My first used pheonixwm, before yabai v 7.0.0 and it was like so:
- toggle yabai floating window
- toggle fullscreen with pheonixwm. (instant)

to go back though, it was messed up. 
I wanted a solution that doesn't use other window managers.
All yabai would have to do is:
- add a new mode besides zoom fullscreen or native fullscreen, such as: toggle-windowed-fullscreen
- make this new fullscreen "mode" ignore yabai window padding and external bar padding. 
This would be an instant fix, which would keep in mind the floating back to bsp tiled.

**Comment by aspauldingcode** [comment]
I have a working 3rd attempt. 
Leaving open for those who want this. It is a very nice feature that should already be there.

**Comment by aspauldingcode** [comment]
This is the functionality I was expecting:
![multi-monitor_windowed-fullscreen_working](https://github.com/koekeishiya/yabai/assets/10196826/1b1848b1-df0b-417d-b31e-cd7ecb4d460f)

**Comment by erezsh** [comment]
> I have a working 3rd attempt.

@aspauldingcode Where did you publish the attempt? Sorry, couldn't find a link

**Comment by koekeishiya** [comment]
When a window enters *native fullscreen*, it changes *presentation mode* which causes various changes to the window. This behaviour is not something that can be activated by yabai. It is special to the "native fullscreen-space" version.

I have added a command that ignores all configured padding (both padding and external_bar padding). yabai is not able to move the window into the Notch region or below the macOS menubar, so that space will always be reserved.
It makes no sense to let the window extend into the Dock area either.

If Dock is hidden/has autohide enabled, yabai will extend the window across the Dock space.
If the MacOS menubar is hidden/autohide enabled, yabai will attempt to place the window at the topmost coordinate (gated by the notch on newer macbooks).

This works for both managed and unmanaged windows.

**Comment by erezsh** [comment]
I tried reinstalling from HEAD. The man page now includes `windowed-fullscreen`, but when I try to run it I get -
```
unknown value 'windowed-fullscreen' given to command '--toggle' for domain 'window'
```

Is there a way to make sure I'm running the HEAD version?

**Comment by aspauldingcode** [comment]
> yabai -m window --toggle windowed-fullscreen

Wow!
I compiled from source and it works. 

Thank you for working on this. However, the problem now is I can drag the window around with the mouse modifier. 

We still need to implement a way to lock/disable mouse-modifier based on if a window is toggled on for a windowed-fullscreen. 

I've done that by using statefiles in my scripts, then setting the at modifier to something ese like fn if I move over the active window. I used sketchybar to check events to calculate this more efficiently. 

It would be great (for scripting purposes like this) if you added a new mouse modifier "none" which effectively disables it manually too. 


Because right now, dragging a window which is meant to be an instant fullscreen is no longer fullscreen when moved around. 

Maybe there could also be a check which sets a window flag to the current window, making sure if a window is-windowed-fullscreen true that the window stays windowed fullscreen until toggled off by the command.

**Comment by erezsh** [comment]
I got it to work too. But it looks just like the zoom-fullscreen feature. i.e. the title-bar is still visible and draggable.

**Comment by koekeishiya** [comment]
> When a window enters native fullscreen, it changes presentation mode which causes various changes to the window. This behaviour is not something that can be activated by yabai. It is special to the "native fullscreen-space" version.

**Comment by erezsh** [comment]
> It makes no sense to let the window extend into the Dock area either.

I would say that it does make sense, if the title bar can't be hidden. But if both are technically impossible, that is another matter.

**Comment by vvatikiotis** [comment]
Is this new option appear in the window properties as a new property when we `--query`, or is it setting the `has-fullscreen-zoom`? I'm on the version 7.1.2 (via homebrew) and toggling `windowed-fullscreen` is not reflected in the window properties.

**Comment by Bellavene** [comment]
Yes, I use this as a workaround

```
#!/bin/zsh
window_info=$(yabai -m query --windows --window | jq -r '.["has-fullscreen-zoom"]')

if [ "$window_info" = "false" ]; then
    yabai -m window --toggle zoom-fullscreen
    sleep 0.15
    yabai -m window --toggle windowed-fullscreen
else
    yabai -m window --toggle windowed-fullscreen
    sleep 0.15
    yabai -m window --toggle zoom-fullscreen
fi
```
